### ISSUE TITLE
MNT: Replace C API usages that return borrowed references

### ISSUE THREAD

**Issue by ngoldbaum** [numpy/numpy#26159]
As noted in [PEP 703](https://peps.python.org/pep-0703/#borrowed-references) C API functions that return borrowed references are problematic in the nogil build.

We need to audit our usages and replace any that upgrade a borrowed reference to an owned reference with alternate C API functions that return new references.

Problematic Functions:

* `PyList_GetItem`
* `PyList_GET_ITEM`
* `PyDict_GetItem`
* `PyDict_GetItemWithError`
* `_PyDict_GetItemStringWithError` (should be replaced anyway because it's a private API)
* `PyDict_GetItemString`

The items in `PyDict` can be replaced using the definitions in `pythoncapi-compat` and I opened https://github.com/numpy/numpy/issues/26158 to discuss adding that project as a submodule. The `PyList` functions don't yet have compat wrappers (see https://github.com/python/pythoncapi-compat/issues/88).

**Comment by arunppsg** [comment]
@ngoldbaum Can I work on replacing usage of `PyList_GET_ITEM`? I intend to replace it with `PyList_GetItemRef` as it returns a strong reference. Is there anything else I should look for?

**Comment by ngoldbaum** [comment]
So, careful, we don’t want to wholesale replace API usages. Sometimes we intentionally create borrowed references. 

Any usages where in the same region of code we first create a borrowed reference and then soon after upgrade it to an owned reference by calling `Py_INCREF` can be replaced by `GetItemRef`. Replacing those usages (and deleting the subsequent `INCREF`) shouldn’t introduce new reference cou ting bugs unless we’re only doing an `INCREF` conditionally.

Basically, be careful, because reference counting bugs can be hard to track down and spot, and we don’t want to introduce any new ones in a refactor.

**Comment by ngoldbaum** [comment]
I think the remaining issue on this task is to add a linter to NumPy's CI to check PRs for new uses of possibly problematic C API items. We should double-check and explicitly whitelist all the remaining ones, then make any new uses get whitelisted as they come in.

I added the sprintable label to this since this should be doable for anyone who understands how to work with github actions and parse git diffs.

**Comment by lvllvl** [comment]
Hi @ngoldbaum, do you mind checking out my PR when you get a chance? I was trying to add a workflow linter for the problematic functions. 

Here's a link to the PR: https://github.com/numpy/numpy/pull/28634
